### PR TITLE
Fix credential leakage in debug logs

### LINKED ISSUE #617: Sensitive credentials logged in debug mode without redaction
## Problem

When debug logging is enabled (`--log-level=debug`), Authorino logs sensitive information in plain text, including:

- OAuth2 client credentials embedded in URLs (via `url.UserPassword()`)
- Bearer tokens in form data during token introspection
- User credentials in Authorization headers
- Session cookies
- API keys
- Kubernetes service account tokens
- **Complete identity objects including Kubernetes Secrets**

### Code Evidence

**OAuth2 Token Introspection** (`pkg/evaluators/identity/oauth2.go`):
```go
tokenIntrospectionURL.User = url.UserPassword(oauth.ClientID, oauth.ClientSecret)
// ...
log.FromContext(ctx).WithName("oauth2").V(1).Info(
    "sending token introspection request",
    "url", tokenIntrospectionURL.String(),  // Logs full URL including credentials
    "data", encodedFormData,                 // Logs form data including token
)
```

**UMA Metadata** (`pkg/evaluators/metadata/uma.go`):
```go
parsedURL.User = url.UserPassword(uma.ClientID, uma.ClientSecret)
// Similar pattern - credentials logged in URL
```

**Identity Validation** (`pkg/service/auth_pipeline.go:243`):
```go
logger.Info("identity validated", "config", conf, "object", extendedObj)
// Logs complete identity object - for API keys, this is the entire Kubernetes Secret
```

**Authorization Pipeline** (`pkg/service/auth_pipeline.go:300`):
```go
logger.Info("evaluating for input", "input", authJSON)
// Logs complete authorization JSON including identity, headers, and metadata
```

**Additional locations:**
- Kubernetes TokenReview requests log the complete token
- Generic HTTP metadata requests and callbacks may contain user credentials
- Incoming authorization requests log all headers including Authorization and Cookie

### Example: API Key Authentication Leakage

When using API key authentication, the entire Kubernetes Secret is logged, exposing:

1. **Base64-encoded secrets** in `data.api_key`: `bmR5QnpyZVV6RjR6cURRc3FTUE1Ia1JocmlFT3RjUng=`
2. **Plaintext secrets** in annotations: `"stringData":{"api_key":"ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx"}`

Example log output:
```json
{
  "level": "debug",
  "msg": "identity validated",
  "object": {
    "kind": "Secret",
    "data": {
      "api_key": "bmR5QnpyZVV6RjR6cURRc3FTUE1Ia1JocmlFT3RjUng="
    },
    "metadata": {
      "annotations": {
        "kubectl.kubernetes.io/last-applied-configuration": "{\"stringData\":{\"api_key\":\"ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx\"}}"
      }
    }
  }
}
```

### Impact

While debug logging is not recommended for production, it is commonly enabled during:
- Troubleshooting in development/staging environments
- Incident response
- Security audits

Credentials leaked to logs can potentially be:
- Stored in centralized logging systems
- Accessed by personnel who shouldn't have credential access
- Exposed if logs are inadvertently shared or compromised

### Affected Components

- OAuth2 token introspection identity verification
- UMA metadata fetching
- Kubernetes TokenReview authentication
- API key authentication (Kubernetes Secrets)
- Generic HTTP metadata and callbacks
- Incoming authorization request logging
- Authorization pipeline evaluation logging
- Identity validation logging

### PR DESCRIPTION
**Pull request by guicassolato**
## Summary

Fixes #617 - Prevents sensitive credentials from leaking into debug logs by implementing comprehensive redaction across all log statements.

## Problem

When debug logging is enabled (`--log-level=debug`), Authorino logged sensitive information in plain text, including:

- OAuth2 client credentials embedded in URLs
- Bearer tokens in form data during token introspection
- User credentials in Authorization headers and cookies
- API keys
- Kubernetes service account tokens
- **Complete Kubernetes Secret objects** (both base64-encoded data and plaintext in annotations)

This exposed credentials in centralized logging systems during troubleshooting, incident response, and security audits.

## Solution

Implemented automatic redaction that replaces sensitive data with `***REDACTED***` while preserving useful debugging information.

### What Gets Redacted

**Default redacted form fields:**
- `token`, `access_token`, `refresh_token`, `id_token`
- `client_secret`, `password`, `secret`
- `api_key`, `apikey`

**Default redacted HTTP headers:**
- `Authorization`, `Cookie`
- `X-API-Key`, `APIKey`

**Additionally redacted:**
- URL credentials (userinfo in URLs)
- Kubernetes TokenReview tokens
- Kubernetes Secret data (all keys in `data` and `stringData`)
- Secrets in `kubectl.kubernetes.io/last-applied-configuration` annotations
- Request bodies (form-encoded fields or complete body)
- Authorization pipeline evaluation data

### Configurable Redaction

New command-line flags allow customization:

```bash
--log-redact-add-field=custom_token          # Add custom sensitive field
--log-redact-remove-field=api_key            # Remove from defaults
--log-redact-add-header=X-Custom-Auth        # Add custom sensitive header
--log-redact-remove-header=cookie            # Remove from defaults
```

### Core Redaction Engine
- **pkg/log/redact.go** - Thread-safe redaction utilities with configurable sensitive fields/headers
- **pkg/log/redact_test.go** - Comprehensive test coverage including Kubernetes Secret redaction

### Log Statement Updates
- **pkg/service/auth_pipeline.go** - Redact identity validation and authorization evaluation logs
- **pkg/evaluators/identity/oauth2.go** - Redact OAuth2 token introspection requests
- **pkg/evaluators/metadata/uma.go** - Redact UMA PAT requests
- **pkg/evaluators/identity/kubernetes_auth.go** - Redact Kubernetes TokenReview tokens
- **pkg/evaluators/metadata/generic_http.go** - Redact generic HTTP metadata/callbacks
- **pkg/service/auth.go** - Redact incoming authorization requests

### Configuration
- **main.go** - Add command-line flags for customizing redaction

### Documentation
- **docs/user-guides/observability.md** - Document redaction feature and configuration

### Prerequisites
```bash
# Set up a local environment based on the current branch
make local-setup FF=1

# Expose the proxy to send requests from localhost
kubectl port-forward deployment/envoy 8000:8000 2>&1 >/dev/null &

# Watch Authorino logs for sensitive data leakage
kubectl logs -f $(kubectl get pods -l authorino-resource=authorino -o name)
```

### Test 1: API Key in Authorization Header

```bash
# Create an API key secret
kubectl apply -f -<<EOF
apiVersion: v1
kind: Secret
metadata:
  name: api-key-1
  labels:
    authorino.kuadrant.io/managed-by: authorino
    group: friends
stringData:
  api_key: ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx
type: Opaque
EOF

# Create AuthConfig expecting API key in Authorization header
kubectl apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta3
kind: AuthConfig
metadata:
  name: talker-api-protection
spec:
  hosts:
  - talker-api.127.0.0.1.nip.io
  authentication:
    "friends":
      apiKey:
        selector:
          matchLabels:
            group: friends
EOF

# Send request and verify logs show ***REDACTED*** instead of the actual API key
curl -H 'Authorization: Bearer ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx' \
  http://talker-api.127.0.0.1.nip.io:8000/hello
```

**Expected in logs:**
- ✅ Authorization header shows `***REDACTED***`
- ✅ Secret `data.api_key` shows `***REDACTED***` (not base64-encoded value)
- ✅ Annotation secrets show `***REDACTED***` (not plaintext value)
- ❌ Should NOT see `ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx`
- ❌ Should NOT see `bmR5QnpyZVV6RjR6cURRc3FTUE1Ia1JocmlFT3RjUng=`

### Test 2: API Key in Custom Header

```bash
# Update AuthConfig to expect API key in custom header
kubectl apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta3
kind: AuthConfig
metadata:
  name: talker-api-protection
spec:
  hosts:
  - talker-api.127.0.0.1.nip.io
  authentication:
    "friends":
      apiKey:
        selector:
          matchLabels:
            group: friends
      credentials:
        customHeader:
          name: APIKEY
EOF

# Send request with custom header and verify redaction
curl -H 'APIKEY: ndyBzreUzF4zqDQsqSPMHkRhriEOtcRx' \
  http://talker-api.127.0.0.1.nip.io:8000/hello
```

**Expected in logs:**
- ✅ Same redaction as Test 1
- ✅ Custom header value is NOT redacted by default (only standard headers)
- ✅ Can be configured with `--log-redact-add-header=APIKEY` if needed

## Notes

- Redaction only applies to debug logs (`V(1)`); info logs already have reduced data exposure
- All tests pass with comprehensive coverage for typed structs, maps, and edge cases
- Thread-safe implementation using `sync.RWMutex` for concurrent access
- Backwards compatible - default behavior improves security without breaking changes

## Summary by CodeRabbit

* **New Features**
  * Debug logging is now safe for production: automatic redaction of sensitive fields, headers, URL credentials, request bodies and auth pipeline data.
  * New CLI flags to customise added/removed sensitive fields and headers (case-insensitive).

* **Documentation**
  * Observability guide updated with redaction behaviour, examples and CLI flag usage.

* **Tests**
  * Added extensive tests covering redaction behaviour and concurrency.

### CODE REVIEW

File: pkg/log/redact_test.go [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+	tests := []struct {
+		name             string
+		url              string
+		shouldContain    string
+		shouldNotContain []string
+	}{
+		{
+			name:             "URL with username and password",
+			url:              "https://clientid:clientsecret@example.com/token",
+			shouldContain:    "@example.com/token",
```
Comment (didierofrivia) [inline pkg/log/redact_test.go:20]: The actual string in the logs would be `"***REDACTED***:***REDACTED***@example.com/token"`, right?

Comment (guicassolato) [reply pkg/log/redact_test.go:20]: I think it's `***REDACTED***:***REDACTED***@example.com/token`.

The test checks that the parts that are not supposed to be redacted stays untouched and the parts that are supposed to be redacted indeed are, all without asserting based on the exact redaction string.

**Review by didierofrivia — APPROVED** [review]
LGTM, tried it and works as expected! 🐸
